File: branches/objectweb/TraceAndPerf/tracelibrary/c++/src/FlushArea.cpp
/* =========================================================================== *
 * This file is part of CARDAMOM (R) which is jointly developed by THALES
 * and SELEX-SI.
 * 
 * It is derivative work based on PERCO Copyright (C) THALES 2000-2003.
 * All rights reserved.
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Library General Public License as published by the
 * Free Software Foundation; either version 2 of the License, or (at your
 * option) any later version.
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Library General
 * Public License along with CARDAMOM; see the file COPYING. If not, write to
 * the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 * =========================================================================== */


#include "tracelibrary/FlushArea.hpp"

#include <algorithm>
#include <sstream>
#include <new>
#include <iostream>


#include "Foundation/common/Assert.hpp"
#include "Foundation/ossupport/OS.hpp"
#include "TraceAndPerf/tracelibrary/FlushAreaMngr.hpp"
#include "tracelibrary/TraceTools.hpp"

namespace Cdmw
{
namespace Trace
{

/**
*Implementation notes: none
*<p>
*Portability issues: none
*<p>
*/



//
// Ctor
//    
FlushArea::FlushArea (size_t nbMessages, size_t sizeArea, size_t sizeOutFlushAreaMsg, unsigned int id)
    throw (OutOfMemoryException)
    : m_messages(nbMessages+1),
      m_id(id),
      m_indexFreeMessage(0),
      m_userSizeArea(sizeArea),
      m_indexFreeBlock(0)
{
	// a supplementary message is allocated in the message container for out flush area error message
	// a supplementary area zone is allocated in raw buffer for error message

	m_totalSizeArea = m_userSizeArea + sizeOutFlushAreaMsg + 100;
    m_pArea = new(std::nothrow) char[m_totalSizeArea];

    if (m_pArea == NULL) 
	{
        CDMW_THROW (OutOfMemoryException);
    }
}




//
// Dtor
//
FlushArea::~FlushArea()
    throw()
{
    delete[] m_pArea;
}



//
// add the specified message
//
FlushArea::AreaStatus FlushArea::add_message (const Message::MessageHeader& header, 
				                              const std::string& body)
    throw()
{

    AreaStatus ret = OK;

    // the actual size of the message body that includes the terminating null character
    size_t actualMessageSize = body.length() + 1;

	// add size to allocate for message header
	actualMessageSize += Message::getMessageHeaderStringSize (header);
    
	// if message is greater than area buffer
    if (actualMessageSize > m_userSizeArea)
    {
        ret = MESSAGE_TOO_BIG;
	}
    else
    {
	    // Is there enough room from available area for user?
	    if ((m_indexFreeBlock + actualMessageSize) <= m_userSizeArea)
		{
		    // We get back the next message from the pool
            Message& message = m_messages[m_indexFreeMessage];

		    // We assign enough room the the internal message
		    //	buffer
		    message.set_buffer (m_pArea + m_indexFreeBlock, actualMessageSize);

            try
            {
                // store message in area buffer
                message.set_message (header, body);

				// increment message index and area buffer index
                m_indexFreeMessage++;
                m_indexFreeBlock += actualMessageSize;

				// if area is full (number of messages reached or area buffer full
                if (m_indexFreeMessage == (m_messages.size()-1) ||
                    m_indexFreeBlock == m_userSizeArea )
                {
                    ret = AREA_FULL;
                }
            }
            catch(const OutOfMemoryException &)
            {
                ret = OUT_OF_MEMORY;
            }
	    }

		// not enough room in area buffer
        else
        {
            ret = SIZE_TOO_SMALL;
	    }
    }	    

    return ret;
}

//
// add the specified out flush area error message
//
void FlushArea::store_outFlushAreaMsg (const std::string& errmsg)
    throw()
{
	// create header for error message
    Message::MessageHeader header;

	header.m_fileLine = 0;
	header.m_fileName = "";
	header.m_userLevel = CdmwTrace::ALL_VALUES;
	header.m_userDomain = CdmwTrace::ALL_DOMAINS;
	header.m_tid = OsSupport::Thread::self();
	header.m_timestamping = OsSupport::OS::get_time();

    // the actual size of the message body that includes the terminating null character
    size_t actualMessageSize = errmsg.length() + 1;

	// add size to allocate for message header
	actualMessageSize += Message::getMessageHeaderStringSize (header);
    
    if (actualMessageSize <= m_totalSizeArea &&
	    m_indexFreeBlock + actualMessageSize <= m_totalSizeArea)
    {
	  // We get back the nex message from the pool
      Message& message = m_messages[m_indexFreeMessage];

	  // We assign enough room the the internal message
	  //	buffer
      message.set_buffer (m_pArea + m_indexFreeBlock, actualMessageSize);

	  // store header and error message
      message.set_message (header, errmsg);

      m_indexFreeMessage++;
      m_indexFreeBlock += actualMessageSize;
    }	    
}

//
//
bool FlushArea::is_empty()
    throw()
{
    return (m_indexFreeMessage == 0);
}


//
// Reset the internal state and delete all messages
//
void FlushArea::reset()
    throw()
{
    CDMW_TRACE_DEBUG(id() << " : FlushArea::reset()");

    m_indexFreeBlock = 0;
    m_indexFreeMessage = 0;
}    


//
// Create the CORBA MessageHeader from the c++ MessageHeader
//
CdmwTrace::MessageHeader* FlushArea::create_MessageHeader (const Trace::Message& message)
    throw (OutOfMemoryException)
{
    using namespace CdmwTrace;

    try 
	{
        // NB: you may ask why this function is not member of MessageHeader,
        // the reason is I don't want to impose in the user namespace Corba symbol
        // or definition, if he doesn't want them (may to port non-Corba code) while
        // keeping the capability to use the trace service.

        const Trace::Message::MessageHeaderRaw& header = message.get_header();

        MessageHeader_var messageHeader= new MessageHeader();

        // Information about the trace origine
        messageHeader->the_file_line = header.m_fileLine;
        char* fileName = CORBA::string_dup (header.m_pFileName);
        check_corba_str (fileName);
        messageHeader->the_file_name = fileName;

        // Information about the user defined characteristic
        messageHeader->the_level.the_value  = header.m_userLevel;
        char* domain = CORBA::string_dup (header.m_pUserDomain);
        check_corba_str (domain);
        messageHeader->the_level.the_domain = domain;

        // Which thread has produced the message
        messageHeader->the_thread_id = header.m_tid;

        // And finally, when the message has been produced (stamping)
        messageHeader->time_stamping.seconds = header.m_timestamping.seconds;
        messageHeader->time_stamping.microseconds = header.m_timestamping.microseconds;

		// return pointer and pass its ownership to caller
        return messageHeader._retn();

    } 
	catch (const std::bad_alloc&)
	{
        CDMW_THROW (OutOfMemoryException);
    }
}



//
// Create the CORBA FormattedMessage from the c++ Message
//
CdmwTrace::FormattedMessage* FlushArea::create_FormattedMessage (const Trace::Message& message)
    throw (OutOfMemoryException)

{
    using namespace CdmwTrace;

    try
	{
        FormattedMessage_var formattedMessage = new FormattedMessage();

        MessageHeader_var messageHeader= create_MessageHeader (message);
        formattedMessage->header = messageHeader.in();
        
        char* buffer = CORBA::string_dup (message.get_bodyBuffer());
        check_corba_str (buffer);
        formattedMessage->message = buffer;

		// return pointer and pass its ownership to caller
        return formattedMessage._retn();
    } 
	catch (const std::bad_alloc&)
	{
        CDMW_THROW (OutOfMemoryException);
    }
}



//
// Create the CORBA sequence of message from the internal message container
//
CdmwTrace::FormattedMessageSeq* FlushArea::create_FormattedMessageSeq()
    throw (OutOfMemoryException)
{
    using namespace CdmwTrace;
    
    try 
	{
        FormattedMessageSeq_var formattedMessageSeq = new FormattedMessageSeq();

        // We create a sequence enough large to hold all contained messages
        formattedMessageSeq->length(m_indexFreeMessage);

        size_t indexMessage(0);

        // We get back each message contained in the flush area
        for (MessageCtr::iterator itrMessage= m_messages.begin(); 
						indexMessage != m_indexFreeMessage; indexMessage++)
        {        
            CDMW_ASSERT (itrMessage != m_messages.end());
            FormattedMessage_var formattedMessage = create_FormattedMessage (*itrMessage);
            formattedMessageSeq[indexMessage] = formattedMessage.in();
            itrMessage++;
        }

		// return pointer and pass its ownership to caller
        return formattedMessageSeq._retn();
        
    } 
	catch (const std::bad_alloc&) 
	{
        CDMW_THROW (OutOfMemoryException);
    }
}




//
// Create from the host information the CORBA header for a set of messages
//
CdmwTrace::MessagesHeader* FlushArea::create_MessagesHeader()
    throw (OutOfMemoryException)
{
    using namespace CdmwTrace;
    
    try 
	{
        MessagesHeader_var messagesHeader= new MessagesHeader();

        // TODO
        messagesHeader->the_host_id = 0;
        char* the_host_name = CORBA::string_dup (OsSupport::OS::get_hostname().c_str());
        check_corba_str (the_host_name);
        messagesHeader->the_host_name = the_host_name;

        messagesHeader->the_process_id = FlushAreaMngr::instance().getProcessId();

        char* the_application_name = CORBA::string_dup (FlushAreaMngr::instance().getApplicationName().c_str());
        check_corba_str (the_application_name);
        messagesHeader->the_application_name = the_application_name;
        
        char* the_process_name = CORBA::string_dup (FlushAreaMngr::instance().getProcessName().c_str());
        check_corba_str (the_process_name);
        messagesHeader->the_process_name = the_process_name;

		// return pointer and pass its ownership to caller
        return messagesHeader._retn();
        
    } 
	catch (const std::bad_alloc&)
	{
        CDMW_THROW (OutOfMemoryException);
    }
}

} // End namespace Trace
} // End namespace Cdmw

File: tags/CARDAMOM_V2_1_CE/TraceAndPerf/tracelibrary/java/src/cdmw/trace/FlushArea.java
/* ===================================================================== */
/*
 * This file is part of CARDAMOM (R) which is jointly developed by THALES
 * and SELEX-SI. It is derivative work based on PERCO Copyright (C) THALES
 * 2000-2003. All rights reserved.
 * 
 * Copyright (C) THALES 2004-2005. All rights reserved
 * 
 * CARDAMOM is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Library General Public License as published
 * by the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * CARDAMOM is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE. See the GNU Library General Public
 * License for more details.
 * 
 * You should have received a copy of the GNU Library General Public
 * License along with CARDAMOM; see the file COPYING. If not, write to the
 * Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
*/
/* ===================================================================== */


package cdmw.trace;


/**
 * This class store an array of Message whose size has been
 * specified during initialisation. When full this area
 * return AREA_FULL when an allocation request is done. 
 * The FlushArea is able to transform all internally stored
 * message in CORBA structure to be sent to Collectors.
 * 
 * Features:
 * No Thread safe 
 *
 * @see FlushAreaMngr
 * @see Message
 *
 */
public class FlushArea {
 
    /**
     * Status of the flush area indicating that a message has been added
     * successfully and the area can receive other messages.
     */
    public static final int OK        = 0;

    /**
     * Status of the flush area indicating that a message has been added
     * successfully but the maximum number of messages has been reached.
     */
    public static final int AREA_FULL = 1;

    /**
     * Status of the flush area indicating that the message cannot 
     * be added to the area as the size of the message exceeds 
     * the available size of the area.
     */
    public static final int SIZE_TOO_SMALL = 2;    

    /**
     * Status of the flush area indicating that the message can never 
     * be added to the area as the size of the message exceeds 
     * the total size of the area.
     */
    public static final int MESSAGE_TOO_BIG = 3;
    
 
    /**
     * Container used to stored the created messages.
     */
    private Message[] messages;
    
    /**
     * Contains the index of the next available message in messages.
     */
    private int indexFreeMessage;
    
    /**
     * Current size of the flush area
     * (in number of characters for messages)
     */
    private int currentSize;

    /**
     * Maximum size of flush area 
     */
    private int maxSize;
        
    /**
     * Id of the flush area (usefull for debug)
     */
    private int id;
 
 
 
    /**
     * Constructs a FlushArea with a maximum size, 
     * a maximum number of messages and no id.
     */
    public FlushArea(int nbMessages, int sizeFlushArea) {
        maxSize = sizeFlushArea;
        currentSize = 0;
        
        // A supplementary message is allocated in the message container 
        // for out flush area error message
        messages = new Message[nbMessages+1];
        for (int i = 0; i < messages.length; i++) {
            messages[i] = new Message();    
        }
        
        indexFreeMessage = 0;
        this.id = 0;
    }
 
 
    /**
     * Constructs a FlushArea with a maximum size, 
     * a maximum number of messages and a specified id.
     */
    public FlushArea(int nbMessages, int sizeFlushArea, int id) {
        this(nbMessages, sizeFlushArea);
        this.id = id;
    }
    
    /**
     * Return the id of the FlushArea. 
     * <br>(usefull for debug)
     */
    public int getId() {
        return id;
    }
    

    /**
     * Adds the specified message to the flush area
     * 
     * @param header the header of the message
     * @param body   the body of the message
     *
     * @return the status of the area after the call (OK or AREA_FULL)
     *
     * @see FlushArea#OK
     * @see FlushArea#AREA_FULL
     */     
    public synchronized int addMessage(MessageHeader header, String body) {
        int messageSize = body.length() + header.threadId.length() +
                          header.fileName.length() + header.userDomain.length() +
                          header.componentName.length(); // ECR-0123
        
        // if message is greater than area buffer
        if (messageSize > maxSize) {

            return MESSAGE_TOO_BIG;
        }
        
        // Is there enough room from available area for user?
        if (messageSize + currentSize > maxSize) {
            return SIZE_TOO_SMALL;
        }
        
        messages[indexFreeMessage].setMessage(header, body);
        indexFreeMessage++;
        currentSize += messageSize;
        
        if ((indexFreeMessage >= messages.length-1) ||
            (currentSize == maxSize))
        {
            return AREA_FULL;
        }
        
        return OK;
    }
    

    /**
     * Returns true if the area is empty.
     */     
    public boolean isEmpty() {
        return indexFreeMessage == 0;
    }


    /**
     * Reset the internal set of the FlushArea, each message
     * stored are lost. This methods is intended to be used when 
     * all contained message have been stored elsewhere
     */ 
    public synchronized void reset() {
        indexFreeMessage = 0;
        currentSize = 0;
        
    }    
    
    
    /**
     * Create the header of one specific message. This
     * function is use internally to build the FormattedMessagesSeq
     * 
     * @param message the message whose header representation must be 
     *   transform into CORBA representation
     *
     * @return the CORBA header
     */ 
    private com.thalesgroup.CdmwTrace.MessageHeader 
    createMessageHeader(Message message) 
    {
        // Information about the user defined characteristic
        com.thalesgroup.CdmwTrace.Trace_Filter level = 
            new com.thalesgroup.CdmwTrace.Trace_Filter(
                    message.getHeader().userLevel,
                    message.getHeader().userDomain,
                    message.getHeader().componentName, // ECR-0123
                    true);
                                                 
        // When the message has been produced (stamping)
        com.thalesgroup.CdmwTrace.Timeval timeval = 
            new com.thalesgroup.CdmwTrace.Timeval(
                    message.getHeader().timeStamping.getSeconds(),
                    message.getHeader().timeStamping.getMicroseconds());
                     
        // Create MessageHeader                                 
        com.thalesgroup.CdmwTrace.MessageHeader messageHeader =
            new com.thalesgroup.CdmwTrace.MessageHeader(level,
                                         // TODO ECR 0050 (add threadName in MessageHeader)
                                         0,
                                         //message.getHeader().threadId,
                                         message.getHeader().fileLine,
                                         message.getHeader().fileName,
                                         timeval);
                           
        return messageHeader;
    }
    
    
    /**
     * Create the CORBA message from a cdmw.trace.Message.
     * This function is use internally to build 
     * the FormattedMessagesSeq
     * 
     * @param message the message to be transformed
     *  into CORBA representation
     *
     * @return the CORBA FormattedMessage
     */ 
    public com.thalesgroup.CdmwTrace.FormattedMessage
    createFormattedMessage(Message message) {
        // get header for current message
        com.thalesgroup.CdmwTrace.MessageHeader messageHeader = 
            createMessageHeader(message);
            
        // create FormattedMessage
        return new com.thalesgroup.CdmwTrace.FormattedMessage(
                        messageHeader,
                        message.getBody());
       
    }
    
    
    
    
    /**
     * Build from the internally stored messages, the
     * equivalent CORBA structure.
     * 
     * @return the CORBA sequence of massages
     */     
    public synchronized com.thalesgroup.CdmwTrace.FormattedMessage[] 
    createFormattedMessageSeq() {
        // Create a sequence with the size of number of messages actually stored
        com.thalesgroup.CdmwTrace.FormattedMessage[] messageSeq =
            new com.thalesgroup.CdmwTrace.FormattedMessage[indexFreeMessage];
            
        // fill the sequence            
        for (int i=0; i < messageSeq.length; i++) {
            // create FormattedMessage
            messageSeq[i] = createFormattedMessage(messages[i]);
                    
        }
        
        // return the sequence
        return messageSeq;
    }
    
    
    
    /**
     * Build from the internally stored messages, the
     * equivalent CORBA Header structure .
     * 
     * @return the CORBA MessagesHeader
     */    
    public com.thalesgroup.CdmwTrace.MessagesHeader createMessagesHeader() {
        com.thalesgroup.CdmwTrace.MessagesHeader messagesHeader = 
            new com.thalesgroup.CdmwTrace.MessagesHeader();
        
        // TODO
        messagesHeader.the_host_id = 0;
        
        // set hostname
        messagesHeader.the_host_name = cdmw.ossupport.OS.getHostname();

        // process id not available with Java
        messagesHeader.the_process_id = 0;
        
        // set application name
        messagesHeader.the_application_name = 
            FlushAreaMngr.getInstance().getApplicationName();
        
        // set process name
        messagesHeader.the_process_name = 
            FlushAreaMngr.getInstance().getProcessName();
        
        return messagesHeader;
    }
    
    
    /**
     * Store the specified out flush area error message
     * 
     * @param errmsg the body of the error message
     */     
    public synchronized void storeOutFlushAreaMsg(String errmsg) {
        MessageHeader messageHeader = 
            new MessageHeader(
                    Thread.currentThread().getName(),
                    com.thalesgroup.CdmwTrace.ALL_COMPONENT_NAMES.value, // ECR-0123
                    com.thalesgroup.CdmwTrace.ALL_DOMAINS.value,
                    com.thalesgroup.CdmwTrace.ALL_VALUES.value,
                    cdmw.ossupport.OS.getTime());
        
        int messageSize = errmsg.length() + messageHeader.threadId.length() +
                          messageHeader.fileName.length() + 
                          messageHeader.userDomain.length() +
                          messageHeader.componentName.length(); // ECR-0123

        messages[indexFreeMessage].setMessage(messageHeader, errmsg);
        indexFreeMessage++;
        
        // Note: error messages doesn't count as user messages for size checking.
        //       Therefore, we doesn't update currentSize.
        
    }
    
    
    
}  

